go/vt/vtgate/scatter_conn.go

```diff
@@ -236,49 +165,72 @@ func (stc *ScatterConn) ExecuteMultiShard(
 		"Execute",
 		rss,
 		session,
-		notInTransaction,
-		func(rs *srvtopo.ResolvedShard, i int, shouldBegin bool, transactionID int64, alias *topodatapb.TabletAlias) (int64, *topodatapb.TabletAlias, error) {
+		func(rs *srvtopo.ResolvedShard, i int, info *shardActionInfo) (*shardActionInfo, error) {
 			var (
 				innerqr *sqltypes.Result
 				err     error
 				opts    *querypb.ExecuteOptions
+				alias   *topodatapb.TabletAlias
 			)
+			transactionID := info.transactionID
+			reservedID := info.reserveID
+
 			if session != nil && session.Session != nil {
 				opts = session.Session.Options
 			}
 
 			switch {
 			case autocommit:
 				// As this is auto-commit, the transactionID is supposed to be zero.
-				if transactionID != 0 {
-					return 0, nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "In autocommit mode, transactionID is non-zero: %d", transactionID)
+				if info.transactionID != int64(0) {
+					return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "in autocommit mode, transactionID should be zero but was: %d", info.transactionID)
 				}
 				innerqr, err = rs.Gateway.Execute(ctx, rs.Target, queries[i].Sql, queries[i].BindVariables, 0, 0, opts)
-			case shouldBegin:
-				innerqr, transactionID, alias, err = rs.Gateway.BeginExecute(ctx, rs.Target, session.Savepoints, queries[i].Sql, queries[i].BindVariables, 0, opts)
-			default:
-				var qs queryservice.QueryService
-				_, usingLegacy := rs.Gateway.(*DiscoveryGateway)
-				if usingLegacy || transactionID == 0 {
-					qs = rs.Gateway
-				} else {
-					qs, err = rs.Gateway.QueryServiceByAlias(alias)
+				if err != nil {
+					return nil, err
 				}
-				if err == nil {
-					innerqr, err = qs.Execute(ctx, rs.Target, queries[i].Sql, queries[i].BindVariables, transactionID, 0, opts)
+			case nothing == info.actionNeeded:
+				qs, err := getQueryService(rs, info)
+				if err != nil {
+					return nil, err
 				}
+				innerqr, err = qs.Execute(ctx, rs.Target, queries[i].Sql, queries[i].BindVariables, info.transactionID, info.reserveID, opts)
+				if err != nil {
+					return nil, err
+				}
+			case begin == info.actionNeeded:
+				qs, err := getQueryService(rs, info)
+				if err != nil {
+					return nil, err
+				}
+				innerqr, transactionID, alias, err = qs.BeginExecute(ctx, rs.Target, session.Savepoints, queries[i].Sql, queries[i].BindVariables, info.reserveID, opts)
+				if err != nil {
+					return info.updateTransactionID(transactionID, alias), err
+				}
+			case reserve == info.actionNeeded:
+				qs, err := getQueryService(rs, info)
+				if err != nil {
+					return nil, err
+				}
+				innerqr, reservedID, alias, err = qs.ReserveExecute(ctx, rs.Target, nil, queries[i].Sql, queries[i].BindVariables, info.transactionID, opts)
+				if err != nil {
+					return info.updateReservedID(reservedID, alias), err
+				}
+			case reserveBegin == info.actionNeeded:
```

Comment:
Is this even possible?

Reply:
Yes. When connecting to a tablet, we could already be in both in a transaction and requiring a reserved connection, and in that case, we would use `ReserveBegin`